sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala

```diff
@@ -95,24 +95,82 @@ package object expressions  {
     }
 
-    // It's possible that `attrs` is a linked list, which can lead to bad O(n) loops when
-    // accessing attributes by their ordinals. To avoid this performance penalty, convert the input
-    // to an array.
-    @transient private lazy val attrsArray = attrs.toArray
+    // Compute min and max expression IDs in a single pass
+    @transient private lazy val minMaxExprId: (Long, Long) = {
+      if (attrs.isEmpty) {
+        (0L, -1L)
+      } else {
+        var min = Long.MaxValue
+        var max = Long.MinValue
+        attrs.foreach { attr =>
+          val id = attr.exprId.id
+          if (id < min) min = id
+          if (id > max) max = id
+        }
+        (min, max)
+      }
+    }
+
+    // Extract as primitive fields to avoid boxing on access
+    @transient private lazy val minExprId: Long = minMaxExprId._1
+    @transient private lazy val maxExprId: Long = minMaxExprId._2
+
+    // Create a directly indexed array with the min and max expression
+    // IDs as an offset.
+    @transient private lazy val ordinalArrays: (Array[Int], Array[Attribute]) = {
+      if (attrs.isEmpty) {
+        (Array.empty[Int], Array.empty[Attribute])
+      } else if (
+        maxExprId - minExprId > Int.MaxValue ||  // prevent overflow
+          maxExprId - minExprId > 2 * attrs.length  // in case of sparse ExprIds
+      ) {
+        (Array.empty[Int], attrs.toArray)
+      } else {
+        // Create directly indexed array
+        val arraySize = (maxExprId - minExprId + 1).toInt
+        val ordinalArray = new Array[Int](arraySize)
+
+        // It's possible that `attrs` is a linked list, which can lead to bad O(n) loops when
+        // accessing attributes by their ordinals. To avoid this performance penalty, convert
+        // the input to an array.
+        val ordinalAttrsArray = new Array[Attribute](attrs.size)
+        java.util.Arrays.fill(ordinalArray, -1) // -1 indicates no attribute with this ID
+
+        // Populate the array based on expression IDs
+        var i = 0
+        attrs.foreach { attr =>
+          val id = attr.exprId.id
+          val arrayIndex = (id - minExprId).toInt
+          ordinalArray(arrayIndex) = i
+          ordinalAttrsArray(i) = attr
+          i += 1
+        }
+
+        (ordinalArray, ordinalAttrsArray)
+      }
+    }
 
-    @transient private lazy val exprIdToOrdinal = {
+    @transient private lazy val exprIdToOrdinal: mutable.LongMap[Int] = {
       val arr = attrsArray
-      val map = Maps.newHashMapWithExpectedSize[ExprId, Int](arr.length)
-      // Iterate over the array in reverse order so that the final map value is the first attribute
-      // with a given expression id.
+
+      // Use LongMap to avoid overhead of HashMap while hashing.
+      val map = new mutable.LongMap[Int](_ => -1, arr.length)
+
+      // Iterate over the array in reverse order so that the final map value is the first
+      // attribute with a given expression id.
```

Comment:
Did we follow this semantic in the new array based approach? i.e. when the input attributes have duplicated ExprIds.